qubes/devices.py

```diff
@@ -115,87 +138,117 @@ class DeviceCollection(object):
-                    device, self._class, self._vm))
+                    device_assignment.ident, self._class, self._vm))
+
+        device = self._device(device_assignment)
         self._vm.fire_event_pre('device-pre-detach:'+self._class, device=device)
-        if persistent:
-            self._set.remove(device)
+        if device in self._set:
+            device_assignment.persistent = True
+            self._set.discard(device_assignment)
+
         self._vm.fire_event('device-detach:' + self._class, device=device)
 
-    def attached(self, persistent=None):
-        '''List devices which are (or may be) attached to this vm
+    def attached(self):
+        '''List devices which are (or may be) attached to this vm '''
+        attached = self._vm.fire_event('device-list-attached:' + self._class)
+        if attached:
+            return [dev for dev, _ in attached]
+
+        return []
+
+    def persistent(self):
+        ''' Devices persistently attached and safe to access before libvirt
+            bootstrap.
+        '''
+        return [self._device(a) for a in self._set]
+
+    def assignments(self, persistent=None):
+        '''List assignments for devices which are (or may be) attached to the
+           vm.
 
         Devices may be attached persistently (so they are included in
         :file:`qubes.xml`) or not. Device can also be in :file:`qubes.xml`,
         but be temporarily detached.
 
-        :param bool persistent: only include devices which are (or are not) \
-        attached persistently - None means both
+        :param bool persistent: only include devices which are or are not
+        attached persistently.
         '''
-        seen = self._set.copy()
-
-        # ask for really attached devices only when requested not only
-        # persistent ones
-        if persistent is not True:
-            attached = self._vm.fire_event(
-                'device-list-attached:' + self._class,
-                persistent=persistent)
-            for device in attached:
-                device_persistent = device in self._set
-                if persistent is not None and device_persistent != persistent:
-                    continue
-                assert device.frontend_domain == self._vm, \
-                    '{!r} != {!r}'.format(device.frontend_domain, self._vm)
-
-                yield device
-
-                try:
-                    seen.remove(device)
-                except KeyError:
-                    pass
-
-        if persistent is False:
-            return
-
-        for device in seen:
-            # get fresh object - may contain updated information
-            device = device.backend_domain.devices[self._class][device.ident]
-            yield device
+
+        devices = self._vm.fire_event('device-list-attached:' + self._class,
+            persistent=persistent)
+        result = []
+        for dev, options in devices:
+            if dev in self._set and persistent is False:
```

Comment:
Booleans should not be checked by comparing identity to ``bool()`` instances, but ``... and persistent`` or `` and not persistent``. What happens if someone uses ``0`` or ``1`` instead of ``True`` and ``False`` for ``persistent``, maybe because they freshly came from linux/driver background and only briefly engage with Python?